File: sortingAlgorithms/day14/quickSort.py
def partition(arr, left, right):
    pivot = arr[right]
    store_index = left-1
    for i in (left, right):
        if arr[i] > pivot:
            store_index += 1
            arr[store_index], arr[i] = arr[i], arr[store_index]

    store_index += 1
    arr[store_index], arr[right] = arr[right], arr[store_index]  
    return  store_index

def quickSort(arr, left, right):
    if(left < right):
        pivot = partition(arr, left, right)
        quickSort(arr, left, pivot)
        quickSort(arr, pivot, right)

if __name__ == "__main__":
    arr = [3, 2, 1, 4, 5]
    quickSort(arr, 0, len(arr) - 1)

    print(arr)

File: sortingAlgorithms/day14/quickSort.cpp
#include <iostream>
#include <vector>
using namespace std;

int partition(vector<int> &arr, int left, int right)
{
    int pivot = arr[right];
    int store_index = left - 1;
    for (int i = 0; i < right; i++)
    {
        if (arr[i] < pivot)
        {
            store_index++;
            swap(arr[store_index], arr[i]);
        }
    }

    swap(arr[store_index + 1], arr[right]);
    return store_index + 1;
}

void quickSort(vector<int> &arr, int left, int right)
{
    int pivot = partition(arr, left, right);
    quickSort(arr, left, pivot);
    quickSort(arr, pivot, right);
}

void printArray(vector<int> &arr)
{
    for (int num : arr)
    {
        cout << num << " ";
    }
    cout << endl;
}

int main()
{
    vector<int> data = {3, 1, 4, 1, 5, 9, 2, 6};
    cout << "原始数组： ";
    printArray(data);
    cout << endl;

    quickSort(data, 0, data.size() - 1);

    cout << "排序后数组：";
    printArray(data);
    cout << endl;
}